Documentation update for Redshift auto discovery support (#9990)
gravitational/teleport: docs/pages/database-access/guides/postgres-redshift.mdx

## Prerequisites

- Teleport version `(=teleport.version=)`.

> - AWS account with Redshift databases and permissions to create and attach IAM

Reviewer: I think "with Redshift databases" can be more precise here. Is this the same as "a Redshift cluster"?

  policies.

## Step 1/5. Install Teleport

(!docs/pages/includes/database-access/start-auth-proxy.mdx!)

## Step 2/5. Create Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

## Step 3/5. Configure IAM

> ### Create IAM policy for Teleport

Reviewer: ```suggestion
### Create an IAM policy for Teleport
```

Teleport needs AWS IAM permissions to be able to:

- Discover and register Redshift databases.
- Manage IAM user or IAM role policies.

> Go to the [Policies](https://console.aws.amazon.com/iamv2/home#/policies) page

Reviewer: ```suggestion
Go to the [Policies](https://console.aws.amazon.com/iamv2/home#/policies) page in the AWS Management Console
```

and create a managed IAM policy for the database agent.

> 

Reviewer: I think there needs to be an introductory sentence or two before the `Tabs` component indicating how a user will know which IAM resource to create.

<Tabs>

>   <TabItem label="Redshift / IAM user">

Reviewer: ```suggestion
  <TabItem label="IAM user">
```

>   Use this policy if your Teleport database agent runs as IAM user (for example,

Reviewer: ```suggestion
  Use this policy if your Teleport database agent runs as an IAM user (for example,
```

  uses AWS credentials file).
  ```json
  {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Action": [
                  "redshift:DescribeClusters",
                  "iam:GetUserPolicy",
                  "iam:PutUserPolicy",
                  "iam:DeleteUserPolicy"
              ],
              "Resource": "*"
          }
      ]
  }
  ```
  </TabItem>
  <TabItem label="Redshift / IAM role">

>   Use this policy if your Teleport database agent runs as IAM role (for example,

Reviewer: ```suggestion
  Use this policy if your Teleport database agent runs as an IAM role (for example,
```

  on an EC2 instance with attached IAM role).
  ```json
  {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Action": [
                  "redshift:DescribeClusters",
                  "iam:GetRolePolicy",
                  "iam:PutRolePolicy",
                  "iam:DeleteRolePolicy"
              ],
              "Resource": "*"
          }
      ]
  }
  ```
  </TabItem>
</Tabs>

> ### Create IAM permission boundary for Teleport

Reviewer: ```suggestion
### Create an IAM permission boundary for Teleport
```

Since Teleport will be managing its own IAM policies for access to Redshift
databases, you need to create a permission boundary to limit its effective
range of permissions.

Create another managed policy that will serve as a permission boundary on the

> same [Policies](https://console.aws.amazon.com/iamv2/home#/policies) page.

Reviewer: ```suggestion
[Policies](https://console.aws.amazon.com/iamv2/home#/policies) page of the AWS Management Console.
```

In addition to the set of permissions you created above, the boundary should

> also include `redshift:GetClusterCredentials` which will grant your Teleport

Reviewer: ```suggestion
also include `redshift:GetClusterCredentials`, which will grant your Teleport
```

> agent the permssion to generate [temporary

Reviewer: ```suggestion
agent the permission to generate [temporary
```

> credentials](https://docs.aws.amazon.com/redshift/latest/mgmt/generating-user-credentials.html)

Reviewer: Do we need the `[temporary credentials]` link here? I think we should prevent the reader from leaving the page as much as possible. We can add this link to the "Next steps" section.

Author: moved to "Next steps". though have to copy content from the included page otherwise the new list item has some bigger gaps from the included list.

to authenticate with Redshift databases.

> <Tabs>

Reviewer: As I say in another comment, I think we can clarify how a reader can determine whether the Teleport database agent runs as an IAM user or IAM role. Currently, each `TabItem` offers examples of each  use case, but not a definitive way to know which one is in effect.

Author: I added some text for this but sure it's clear enough. please review and advise. thanks!

>   <TabItem label="Redshift / IAM user">

Reviewer: ```suggestion
  <TabItem label="IAM user">
```

>   Use this permission boundary if your Teleport database agent runs as IAM user

Reviewer: ```suggestion
  Use this permission boundary if your Teleport database agent runs as an IAM user
```

  (for example, uses AWS credentials file).
  ```json
  {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Action": [
                  "redshift:DescribeClusters",
                  "redshift:GetClusterCredentials",
                  "iam:GetUserPolicy",
                  "iam:PutUserPolicy",
                  "iam:DeleteUserPolicy"
              ],
              "Resource": "*"
          }
      ]
  }
  ```
  </TabItem>

>   <TabItem label="Redshift / IAM role">

Reviewer: ```suggestion
  <TabItem label="IAM role">
```

>   Use this permission boundary if your Teleport database agent runs as IAM role

Reviewer: ```suggestion
  Use this permission boundary if your Teleport database agent runs as an IAM role
```

  (for example, on an EC2 instance with attached IAM role).
  ```json
  {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Action": [
                  "redshift:DescribeClusters",
                  "redshift:GetClusterCredentials",
                  "iam:GetRolePolicy",
                  "iam:PutRolePolicy",
                  "iam:DeleteRolePolicy"
              ],
              "Resource": "*"
          }
      ]
  }
  ```
  </TabItem>
</Tabs>

> ### Attach policy and boundary to IAM identity

Reviewer: ```suggestion
### Attach the policy and boundary to an IAM identity
```

> Attach created policy and permission boundary to the IAM identity your Teleport

Reviewer: ```suggestion
Attach the policy and permission boundary you created earlier to the IAM identity your Teleport
```

> database agent will be using. For example, if the agent runs as an IAM user:

Reviewer: Maybe it's my end-of-the-day brain, but I think we need a sentence explaining what's going on in this image.

<Figure>
![IAM user](../../../img/database-access/iam@2x.png)
</Figure>

<Admonition type="note" title="Self-managed IAM">
  If you prefer to self-manage IAM for your Redshift databases, see [AWS
  reference](../reference/aws.mdx) for details.
</Admonition>

> ## Step 4/5. Start database agent

Reviewer: ```suggestion
## Step 4/5. Start the database agent
```

    tags: # AWS database resource tags to match
      "*": "*"
```

Start the database agent:

```code
$ teleport start --config=/etc/teleport.yaml
```

The agent will discover all Redshift databases according to the configuration
and register them in the cluster. The agent will also attempt to configure IAM
access policies for the discovered databases. Keep in mind that AWS IAM changes
may not propagate immediately and can take a few minutes to come into effect.

<Admonition type="note" title="AWS credentials">
  Teleport database agent uses the default credential provider chain to find AWS
  credentials. See [Specifying Credentials](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials)
  for more information.
</Admonition>

## Step 5/5. Connect

Once the database agent has started and joined the cluster, login to see the
registered databases:

```code
$ tsh login --proxy=teleport.example.com --user=alice
$ tsh db ls
# Name        Description                    Labels
# ----------- ------------------------------ --------
# my-redshift Redshift cluster in us-east-1  ...
```

> Log into particular database using `tsh db login` command:

Reviewer: ```suggestion
Log into a particular database using the `tsh db login` command:
```

```code
$ tsh db login my-redshift
```

<Admonition type="tip" title="Tip">
  You can be logged into multiple databases simultaneously.
</Admonition>

You can optionally specify the database name and the user to use by default
when connecting to the database instance:

```code
$ tsh db login --db-user=awsuser --db-name=dev my-redshift
```

> <Admonition type="note" title="Users">

Reviewer: Can we make the title more specific than "Users"?

```yaml
teleport:
  data_dir: /var/lib/teleport
  auth_token: /tmp/token
  auth_servers:
  - teleport.example.com:443 # Teleport proxy address to connect to
auth_service:
  enabled: "no"
proxy_service:
  enabled: "no"
db_service:
  enabled: "yes"
  aws: # Matchers for registering AWS-hosted databases.
  - types: ["redshift"]
    regions: ["us-west-1"] # AWS regions to fetch databases from
    tags: # AWS database resource tags to match
      "*": "*"
```

Start the database agent:

```code
$ teleport start --config=/etc/teleport.yaml
```

The agent will discover all Redshift databases according to the configuration
and register them in the cluster. The agent will also attempt to configure IAM
access policies for the discovered databases. Keep in mind that AWS IAM changes
may not propagate immediately and can take a few minutes to come into effect.

<Admonition type="note" title="AWS credentials">
  Teleport database agent uses the default credential provider chain to find AWS
  credentials. See [Specifying Credentials](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials)
  for more information.
</Admonition>

## Step 5/5. Connect

> Once the database agent has started and joined the cluster, login to see the

Reviewer: ```suggestion
Once the database agent has started and joined the cluster, log in to see the
```

To log out of the database and remove credentials:

```code
$ tsh db logout my-redshift
```

## Next steps

- Learn more about [using IAM authentication to generate database user
  credentials](https://docs.aws.amazon.com/redshift/latest/mgmt/generating-user-credentials.html) for AWS Redshift.
- Learn how to [restrict access](../rbac.mdx) to certain users and databases.

> - View [High Availability (HA)](../guides/ha.mdx) guide.

Reviewer: ```suggestion
- View the [High Availability (HA)](../guides/ha.mdx) guide.
```

> - Take a look at YAML configuration [reference](../reference/configuration.mdx).

Reviewer: ```suggestion
- Take a look at the YAML configuration [reference](../reference/configuration.mdx).
```
